File: iam454/week8/leet1760.js
/**
 * @param {number[]} nums
 * @param {number} maxOperations
 * @return {number}
 */
var minimumSize = function (nums, maxOperations) {
  const canDivide = (penalty) => {
    let totalOps = 0;
    for (const num of nums) {
      totalOps += Math.ceil(num / penalty) - 1;
      if (totalOps > maxOperations) {
        return false;
      }
    }

    return true;
  };

  let l = 1;
  let r = Math.max(...nums);
  let ans = 0;

  while (l <= r) {
    let mid = Math.floor((l + r) / 2);
    if (canDivide(mid)) {
      ans = mid;
      r = mid - 1;
    } else {
      l = mid + 1;
    }
  }

  return ans;
};

File: fru1tworld/week06~10/week08/leet1760/leet1760.cpp
class Solution {
public:
    int minimumSize(vector<int>& nums, int maxOperations) {
        int len = nums.size();
        int l = 1 ,r = *max_element(nums.begin(), nums.end()); 
        int ans = r; 
        while(l+1 < r){
            int m = (l+r)/2; 
            int cnt = 0;
            for(int i=0; i<len; i++){
                if(m >= nums[i]) continue; 
                else{
                    cnt += (nums[i]-1)/m;
                }
            }
            if(cnt > maxOperations){
                l = m; 
            }else{
                r = m; 
                ans = r; 
            }
        }
        return ans; 
    }
};